File: python/open_guardrail/guards/command_hierarchy.py
"""Prevents lower-priority instructions from overriding higher ones."""
from __future__ import annotations

import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_OVERRIDE_PATTERNS = [
    "ignore previous",
    "disregard above",
    "override instructions",
    "new instructions:",
    "forget everything",
    "ignore all prior",
    "skip the rules",
    "bypass restrictions",
    "you are now",
    "act as if",
    "pretend that",
    "your new role",
]


class _CommandHierarchy:
    def __init__(self, *, action: str = "block") -> None:
        self.name = "command-hierarchy"
        self.action = action

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        lower = text.lower()
        matched = [p for p in _OVERRIDE_PATTERNS if p in lower]
        triggered = len(matched) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="command-hierarchy",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message="Command hierarchy override attempt detected" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"matched_patterns": matched} if triggered else None,
        )


def command_hierarchy(*, action: str = "block") -> _CommandHierarchy:
    return _CommandHierarchy(action=action)

File: packages/guards/src/command-hierarchy.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface CommandHierarchyOptions {
  action: 'block' | 'warn';
}

const OVERRIDE_PATTERNS = [
  'ignore previous',
  'disregard above',
  'override instructions',
  'new instructions:',
  'forget everything',
  'ignore all prior',
  'skip the rules',
  'bypass restrictions',
  'you are now',
  'act as if',
  'pretend that',
  'your new role',
];

export function commandHierarchy(options: CommandHierarchyOptions): Guard {
  return {
    name: 'command-hierarchy',
    version: '0.1.0',
    description: 'Prevents lower-priority instructions from overriding higher ones',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const lower = text.toLowerCase();
      const matched = OVERRIDE_PATTERNS.filter((p) => lower.includes(p));
      const triggered = matched.length > 0;
      return {
        guardName: 'command-hierarchy',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { matchedPatterns: matched }
          : undefined,
      };
    },
  };
}
